lib/chef-dk/completions/bash.sh.erb

```diff
@@ -1,5 +1,73 @@
+# Chef tab-completion  v1.1
+# Created by Jose Luis Mantilla 2018
+# email: joseluismantilla@gmail.com
+#
+# Features
+# chef command tab-completion 
+# knife command filled out dynamically the first time according to the chefdk version and caching 
+# in the home directory for improving speed.
+
+
 _chef_comp() {
-    local COMMANDS="<%= commands.keys.join(' ')-%>"
-    COMPREPLY=($(compgen -W "$COMMANDS" -- ${COMP_WORDS[COMP_CWORD]} ))
+    local cur prev opts 
+    COMPREPLY=()
+    cur="${COMP_WORDS[COMP_CWORD]}"
+    prev="${COMP_WORDS[COMP_CWORD-1]}"
+
+    if [[ $prev == "chef" ]]; then
+        local COMMANDS="<%= commands.keys.join(' ')-%>"
+        COMPREPLY=($(compgen -W "$COMMANDS" -- ${COMP_WORDS[COMP_CWORD]} ))
+    else
+        case "${prev}" in
+        gem)
+            opts="install list build help server"
+            COMPREPLY=( $(compgen -W "$opts" ${cur}) )
+            ;;
+        generate )
+            local running="app attribute Available cookbook file generator helpers lwrp policyfile recipe repo resource template"
+            COMPREPLY=( $(compgen -W "${running}" -- ${cur}) )
+            return 0
+            ;;
+        *)
+            ;;
+        esac
+    fi
+
 }
+
+_knife_completion() 
+{
+    local cur prev opts base
+    COMPREPLY=()
+    cur="${COMP_WORDS[COMP_CWORD]}"
+    prev="${COMP_WORDS[COMP_CWORD-1]}"
+    file_caching="/home/$USER/.knife.tmp"
```

Comment:
Could we do `$HOME/.knife.tmp` so this would also work if the home dir is not `/home`? IE, MacOS

Reply:
Also seems like we could put it in `$HOME/.chef/.knife_completions` or something to keep it next to other chef/chefdk configurations.